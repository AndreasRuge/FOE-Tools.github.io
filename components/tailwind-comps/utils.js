import styleProps from "./props";

/**
 * Checks if object has a specific property.
 * @param {Object} obj
 * @param {String} prop
 */
export const hasOwn = (obj, prop) => Object.prototype.hasOwnProperty.call(obj, prop);

/**
 * Checks if a value is defined
 * @param {*} v
 * @returns {Boolean}
 */
export function isDef(v) {
  return v !== undefined && v !== null;
}

export function useId(size = 3) {
  let uuid = "";
  const dictionary = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
  for (let i = 0; i < size; i++) {
    uuid += dictionary.charAt(Math.floor(Math.random() * dictionary.length));
  }
  return uuid;
}

/**
 * Queries an element from the DOM
 * @param {String} selector Element selector
 * @param {Node} domain HTML element in which to query for element
 * @returns {Node} Node
 */
export const getElement = (selector, domain) => {
  if (!domain) {
    return document.querySelector(selector);
  } else {
    return domain.querySelector(selector);
  }
};

/**
 * @description Wraps and executes both user and internal event handlers for a single event
 * @param {Function} theirHandler Userland event handler
 * @param {*} ourHandler Internal Vue chakra event handler
 */
export const wrapEvent = (theirHandler, ourHandler) => (event) => {
  if (theirHandler) {
    theirHandler(event);
  }

  if (!event.defaultPrevented) {
    return ourHandler(event);
  }
};

/**
 * Converts a kebab-case string into camel case
 * @param {String} string
 */
export function camelize(string) {
  return string.replace(/[.-](\w|$)/g, function (_, x) {
    return x.toUpperCase();
  });
}

/** Filter attrs and return object of chakra props */
export function extractChakraAttrs(attrs) {
  const styleAttrs = {};
  const nativeAttrs = {};

  for (const _prop in attrs) {
    const prop = camelize(_prop);
    if (styleProps[prop]) {
      styleAttrs[prop] = attrs[_prop];
    } else {
      nativeAttrs[_prop] = attrs[_prop];
    }
  }
  return { styleAttrs, nativeAttrs };
}

/**
 * Makes a cache watcher handler for data property.
 * This utility helps prevent unnecessary re-renders
 * for primitives with changes in the $parent $attrs
 * and $listeners objects
 * @param {String} property
 */
export function createWatcher(property) {
  return {
    handler(newVal, oldVal) {
      for (const key in oldVal) {
        if (!hasOwn(newVal, key)) {
          this.$delete(this.$data[property], key);
        }
      }
      for (const key in newVal) {
        this.$set(this.$data[property], key, newVal[key]);
      }
    },
    immediate: true,
  };
}

/**
 * Create mixin for style attributes
 * @param {String} name Component name
 */
export const createStyledAttrsMixin = (name) => ({
  name,
  inheritAttrs: false,
  data() {
    return {
      attrs$: {},
      listeners$: {},
    };
  },
  props: {
    as: {
      type: [String, Object],
      default: "div",
    },
    to: String,
  },
  computed: {
    /** Split style attributes and native attributes */
    splitProps() {
      const $attrs = this.$data.attrs$;
      const styles = Object.assign({}, this.componentStyles || {}, $attrs);

      const { styleAttrs, nativeAttrs } = extractChakraAttrs(styles);
      return {
        styleAttrs,
        nativeAttrs,
      };
    },
    /** Computed attributes object */
    computedAttrs() {
      return {
        ...this.splitProps.nativeAttrs,
      };
    },
    /** Computed listeners object */
    computedListeners() {
      return this.$data.listeners$;
    },
  },
  created() {
    this.$watch("$attrs", createWatcher("attrs$"));
    this.$watch("$listeners", createWatcher("listeners$"));
  },
});

/**
 * Clones a single VNode
 * @param {Vue.VNode} vnodes VNode to be cloned
 * @param {Function} createElement Render function
 * @returns {Vue.VNode} Cloned VNodes
 */
export function cloneVNode(vnode, createElement) {
  const clonedChildren = vnode.children && vnode.children.map((vnode) => cloneVNode(vnode, createElement));
  const cloned = createElement(vnode.tag, vnode.data, clonedChildren);
  cloned.text = vnode.text;
  cloned.isComment = vnode.isComment;
  cloned.componentOptions = vnode.componentOptions;
  cloned.fnOptions = vnode.fnOptions;
  cloned.fnContext = vnode.fnContext;
  cloned.elm = vnode.elm;
  cloned.context = vnode.context;
  cloned.ns = vnode.ns;
  cloned.isStatic = vnode.isStatic;
  cloned.key = vnode.key;
  cloned.data = vnode.data;
  return cloned;
}

/**
 * Clones an array of VNodes
 * @param {Array<Vue.VNode>} vnodes Array of VNodes to be cloned
 * @param {Function} createElement Render function
 * @returns {Array<Vue.VNode>} Cloned VNodes Array
 */
export function cloneVNodes(vnodes, createElement) {
  const clonedVNodes = vnodes.map((vnode) => cloneVNode(vnode, createElement));
  return clonedVNodes;
}

/**
 * Clones VNode with merged data
 * @param {Vue.VNode} vnode VNode
 * @param {Object} data VNode data
 * @param {Function} h Render function
 */
export function cloneVNodeElement(vnode, { props, attrs, children, ...rest }, h) {
  const cloned = cloneVNode(vnode, h);
  return h(
    cloned.componentOptions.Ctor,
    {
      ...cloned.data,
      ...(cloned.componentOptions.listeners || {}),
      props: {
        ...(cloned.data.props || {}),
        ...cloned.componentOptions.propsData,
        ...props,
      },
      attrs: {
        ...(cloned.data.attrs || {}),
        ...attrs,
      },
      ...rest,
    },
    cloned.componentOptions.children || children
  );
}

/**
 * Checks whether a vnode is an async placeholder
 * @param {Vue.VNode} node
 * @returns {Boolean}
 */
function isAsyncPlaceholder(node) {
  return node.isComment && node.asyncFactory;
}

/**
 * Get's the first child component VNode from an array of VNodes
 * @param {Array<Vue.VNode>} children
 * @returns {Vue.VNode}
 */
export function getFirstComponentChild(children) {
  if (Array.isArray(children)) {
    for (let i = 0; i < children.length; i++) {
      const c = children[i];
      if (isDef(c) && (isDef(c.componentOptions) || isAsyncPlaceholder(c))) {
        return c;
      }
    }
  }
}

/**
 * Clears out all undefined or nullish vnodes
 * @param {Array<Vue.VNode>} vnodes Array of VNodes
 * @returns {Array<Vue.VNode>}
 */
export function cleanChildren(vnodes) {
  if (!vnodes) return [];
  return vnodes.filter((vnode) => vnode.tag);
}

/**
 * Extracts native and nonNative event handlers for functional components
 *
 * The returned object returns `native` object contains listeners that
 * MUST be passed to a Vue component in the render function.
 *
 * The returned `nonNative` object contains other native listeners that
 * can be passed to a native HTML element in the render function.
 *
 * @param {Object} context Render function context
 * @param {Object} nonNative Object of VNode `on` event handlers
 * @returns {{ nonNative: Object<String, Function>, native: Object<String, Function>}}
 *
 * @example
 * import Comp from 'comp'
 *
 * const newComp = {
 *  functional: true,
 *  render(h, context) {
 *    const { native, nonNative } = extractListeners(context, componentEvents)
 *    return h(Comp, {
 *      ...context.data
 *      on: nonNative,
 *      nativeOn: native,
 *    }, context.slots ? context.slots().default : context.children)
 *  }
 * }
 */
export const extractListeners = (context, nonNative = {}) => {
  const { listeners } = context;
  const native = {};

  for (const listener in listeners) {
    if (!nonNative[listener]) {
      native[listener] = listeners[listener];
    }
  }

  return { native, nonNative };
};
