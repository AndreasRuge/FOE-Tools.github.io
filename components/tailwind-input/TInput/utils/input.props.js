export const inputProps = {
  as: {
    type: String,
    default: "input",
  },
  _ariaLabel: String,
  _ariaDescribedby: String,
  isFullWidth: {
    type: Boolean,
    default: true,
  },
  focusStyle: {
    type: Boolean,
    default: true,
  },
  isReadOnly: Boolean,
  isDisabled: Boolean,
  isInvalid: Boolean,
  isRequired: Boolean,
  inputId: String,
  rows: Number,
  autoComplete: String,
  value: {
    type: [String, Number],
    default: undefined,
  },
};
