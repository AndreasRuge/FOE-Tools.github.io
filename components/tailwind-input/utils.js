import styleProps from "./props";

/**
 * Checks if object has a specific property.
 * @param {Object} obj
 * @param {String} prop
 */
export const hasOwn = (obj, prop) => Object.prototype.hasOwnProperty.call(obj, prop);

/**
 * Checks if a value is defined
 * @param {*} v
 * @returns {Boolean}
 */
export function isDef(v) {
  return v !== undefined && v !== null;
}

export function useId(size = 3) {
  let uuid = "";
  const dictionary = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
  for (let i = 0; i < size; i++) {
    uuid += dictionary.charAt(Math.floor(Math.random() * dictionary.length));
  }
  return uuid;
}

/**
 * Queries an element from the DOM
 * @param {String} selector Element selector
 * @param {Node} domain HTML element in which to query for element
 * @returns {Node} Node
 */
export const getElement = (selector, domain) => {
  if (!domain) {
    return document.querySelector(selector);
  } else {
    return domain.querySelector(selector);
  }
};

/**
 * @description Wraps and executes both user and internal event handlers for a single event
 * @param {Function} theirHandler Userland event handler
 * @param {*} ourHandler Internal Vue chakra event handler
 */
export const wrapEvent = (theirHandler, ourHandler) => (event) => {
  if (theirHandler) {
    theirHandler(event);
  }

  if (!event.defaultPrevented) {
    return ourHandler(event);
  }
};

/**
 * Converts a kebab-case string into camel case
 * @param {String} string
 */
export function camelize(string) {
  return string.replace(/[.-](\w|$)/g, function (_, x) {
    return x.toUpperCase();
  });
}

/** Filter attrs and return object of chakra props */
export function extractChakraAttrs(attrs) {
  const styleAttrs = {};
  const nativeAttrs = {};

  for (const _prop in attrs) {
    const prop = camelize(_prop);
    if (styleProps[prop]) {
      styleAttrs[prop] = attrs[_prop];
    } else {
      nativeAttrs[_prop] = attrs[_prop];
    }
  }
  return { styleAttrs, nativeAttrs };
}

/**
 * Makes a cache watcher handler for data property.
 * This utility helps prevent unnecessary re-renders
 * for primitives with changes in the $parent $attrs
 * and $listeners objects
 * @param {String} property
 */
export function createWatcher(property) {
  return {
    handler(newVal, oldVal) {
      for (const key in oldVal) {
        if (!hasOwn(newVal, key)) {
          this.$delete(this.$data[property], key);
        }
      }
      for (const key in newVal) {
        this.$set(this.$data[property], key, newVal[key]);
      }
    },
    immediate: true,
  };
}

/**
 * Create mixin for style attributes
 * @param {String} name Component name
 */
export const createStyledAttrsMixin = (name) => ({
  name,
  inheritAttrs: false,
  data() {
    return {
      attrs$: {},
      listeners$: {},
    };
  },
  props: {
    as: {
      type: [String, Object],
      default: "div",
    },
    to: String,
  },
  computed: {
    /** Split style attributes and native attributes */
    splitProps() {
      const $attrs = this.$data.attrs$;
      const styles = Object.assign({}, this.componentStyles || {}, $attrs);

      const { styleAttrs, nativeAttrs } = extractChakraAttrs(styles);
      return {
        styleAttrs,
        nativeAttrs,
      };
    },
    /** Computed attributes object */
    computedAttrs() {
      return {
        ...this.splitProps.nativeAttrs,
      };
    },
    /** Computed listeners object */
    computedListeners() {
      return this.$data.listeners$;
    },
  },
  created() {
    this.$watch("$attrs", createWatcher("attrs$"));
    this.$watch("$listeners", createWatcher("listeners$"));
  },
});
