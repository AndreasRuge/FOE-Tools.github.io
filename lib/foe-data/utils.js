/**
 * Rounds one number to the nearest integer multiple of another
 *
 * @param value {number} The number to round to the nearest integer multiple of another
 * @param factor {number} The number to whose multiples value will be rounded
 * @returns {number} Return the rounded value
 */
function mround(value, factor) {
  return Math.round(value / factor) * factor;
}

/**
 * Generate reward for all of the first 5 places
 *
 * @param firstRankReward {number} Reward of the first place
 * @returns {*[]} Return an array of number
 */
function generateReward(firstRankReward) {
  let result = [firstRankReward];

  result[result.length] = mround(result[result.length - 1] / 2, 5);
  result[result.length] = mround(result[result.length - 1] / 3, 5);
  result[result.length] = mround(result[result.length - 1] / 4, 5);
  result[result.length] = mround(result[result.length - 1] / 5, 5);

  return result;
}

/**
 * Default Ages GB cost (for no age and HMA)
 * @type {*[]} Return array of cost that contains an object with:
 * - cost {number}: level cost
 * - reward {Array}: Array of 5 integer value that represent the reward (in FP) of the first five places
 */
const defaultCost = [
  { cost: 650, reward: generateReward(65) },
  { cost: 667, reward: generateReward(75) },
  { cost: 683, reward: generateReward(85) },
  { cost: 700, reward: generateReward(95) },
  { cost: 718, reward: generateReward(100) },
  { cost: 736, reward: generateReward(110) },
  { cost: 754, reward: generateReward(120) },
  { cost: 773, reward: generateReward(130) },
  { cost: 792, reward: generateReward(140) },
  { cost: 812, reward: generateReward(150) },
  { cost: 833, reward: generateReward(155) },
  { cost: 853, reward: generateReward(165) },
  { cost: 875, reward: generateReward(175) },
  { cost: 897, reward: generateReward(185) },
  { cost: 919, reward: generateReward(195) },
  { cost: 942, reward: generateReward(205) },
  { cost: 965, reward: generateReward(215) },
  { cost: 990, reward: generateReward(225) },
  { cost: 1014, reward: generateReward(235) },
  { cost: 1040, reward: generateReward(245) },
  { cost: 1066, reward: generateReward(255) },
  { cost: 1092, reward: generateReward(265) },
  { cost: 1120, reward: generateReward(275) },
  { cost: 1147, reward: generateReward(285) },
  { cost: 1176, reward: generateReward(300) },
  { cost: 1206, reward: generateReward(310) },
  { cost: 1236, reward: generateReward(320) },
  { cost: 1267, reward: generateReward(330) },
  { cost: 1298, reward: generateReward(340) },
  { cost: 1331, reward: generateReward(350) },
  { cost: 1364, reward: generateReward(365) },
  { cost: 1398, reward: generateReward(375) },
  { cost: 1433, reward: generateReward(385) },
  { cost: 1469, reward: generateReward(395) },
  { cost: 1505, reward: generateReward(405) },
  { cost: 1543, reward: generateReward(420) },
  { cost: 1582, reward: generateReward(430) },
  { cost: 1621, reward: generateReward(440) },
  { cost: 1662, reward: generateReward(450) },
  { cost: 1703, reward: generateReward(465) },
  { cost: 1746, reward: generateReward(475) },
  { cost: 1789, reward: generateReward(495) },
  { cost: 1834, reward: generateReward(500) },
  { cost: 1880, reward: generateReward(510) },
  { cost: 1927, reward: generateReward(520) },
  { cost: 1975, reward: generateReward(535) },
  { cost: 2025, reward: generateReward(545) },
  { cost: 2075, reward: generateReward(555) },
  { cost: 2127, reward: generateReward(570) },
  { cost: 2180, reward: generateReward(580) },
  { cost: 2235, reward: generateReward(590) },
  { cost: 2290, reward: generateReward(605) },
  { cost: 2348, reward: generateReward(615) },
  { cost: 2406, reward: generateReward(630) },
  { cost: 2467, reward: generateReward(640) },
  { cost: 2528, reward: generateReward(650) },
  { cost: 2591, reward: generateReward(665) },
  { cost: 2656, reward: generateReward(675) },
  { cost: 2723, reward: generateReward(690) },
  { cost: 2791, reward: generateReward(700) },
  { cost: 2860, reward: generateReward(715) },
  { cost: 2932, reward: generateReward(725) },
  { cost: 3005, reward: generateReward(735) },
  { cost: 3080, reward: generateReward(750) },
  { cost: 3157, reward: generateReward(760) },
  { cost: 3236, reward: generateReward(775) },
  { cost: 3317, reward: generateReward(785) },
  { cost: 3400, reward: generateReward(800) },
  { cost: 3485, reward: generateReward(810) },
  { cost: 3572, reward: generateReward(825) },
  { cost: 3661, reward: generateReward(835) },
  { cost: 3753, reward: generateReward(850) },
  { cost: 3847, reward: generateReward(860) }
];

module.exports = {
  mround: mround,
  generateReward: generateReward,
  defaultCost: defaultCost
};
