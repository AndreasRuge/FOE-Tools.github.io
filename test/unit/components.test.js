import { doTestOf } from "./utils";

describe.skip("Components", () => {
  doTestOf("components");
});
