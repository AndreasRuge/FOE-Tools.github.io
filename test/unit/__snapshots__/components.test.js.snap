// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components GbInvestment Call "addInvestor" 1`] = `
Array [
  Object {
    "isPotentialSniper": true,
    "value": 1,
  },
]
`;

exports[`Components GbInvestment Call "addInvestor" with big value 1`] = `
Array [
  Object {
    "isPotentialSniper": true,
    "value": 326,
  },
]
`;

exports[`Components GbInvestment Call "calculate" with maxInvestment < 0 1`] = `
Object {
  "cost": 1210,
  "investment": Array [
    Object {
      "cumulativeInvestment": 1110,
      "defaultParticipationIndex": 0,
      "expectedParticipation": 238,
      "isInvestorParticipation": true,
      "participation": 500,
      "preparation": 500,
      "reward": 125,
      "roi": -262,
      "snipe": Object {
        "fp": -1,
        "roi": 0,
      },
    },
    Object {
      "cumulativeInvestment": 1120,
      "defaultParticipationIndex": 1,
      "expectedParticipation": 124,
      "isInvestorParticipation": true,
      "participation": 100,
      "preparation": 510,
      "reward": 65,
      "roi": 24,
      "snipe": Object {
        "fp": -1,
        "roi": 0,
      },
    },
    Object {
      "cumulativeInvestment": 1182,
      "defaultParticipationIndex": -1,
      "expectedParticipation": 38,
      "isInvestorParticipation": false,
      "participation": 38,
      "preparation": 534,
      "reward": 20,
      "roi": 0,
      "snipe": Object {
        "fp": 50,
        "roi": -12,
      },
    },
    Object {
      "cumulativeInvestment": 1200,
      "defaultParticipationIndex": 2,
      "expectedParticipation": 10,
      "isInvestorParticipation": true,
      "participation": 10,
      "preparation": 552,
      "reward": 5,
      "roi": 0,
      "snipe": Object {
        "fp": 55,
        "roi": -45,
      },
    },
    Object {
      "defaultParticipationIndex": -1,
      "expectedParticipation": 0,
      "isInvestorParticipation": false,
      "preparation": 562,
      "reward": 0,
      "roi": 0,
      "snipe": Object {
        "fp": 0,
        "roi": 0,
      },
    },
  ],
  "level": 10,
  "otherInvestment": Array [],
  "totalPreparations": 562,
}
`;

exports[`Components GbInvestment Call "changePlaceFree" with invalid data 1`] = `
Array [
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P2(67) P3(19) P4(10)",
  },
  Object {
    "active": false,
    "message": "P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2 > 10",
  },
]
`;

exports[`Components GbInvestment Call "removeInvestor" with invalid index 1`] = `
Array [
  Object {
    "isPotentialSniper": true,
    "value": 326,
  },
  Object {
    "isPotentialSniper": true,
    "value": 5,
  },
]
`;

exports[`Components GbInvestment Call "removeInvestor" with invalid index 2`] = `
Array [
  Object {
    "isPotentialSniper": true,
    "value": 326,
  },
  Object {
    "isPotentialSniper": true,
    "value": 5,
  },
]
`;

exports[`Components GbInvestment Call 'addPromotionMessageTemplate' 'templateToAdd' not set 1`] = `
Array [
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2,1 > 10",
  },
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
]
`;

exports[`Components GbInvestment Call 'addPromotionMessageTemplate' 'templateToAdd' not set 2`] = `
Array [
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2,1 > 10",
  },
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
]
`;

exports[`Components GbInvestment Call 'addPromotionMessageTemplate' 1`] = `
Array [
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2,1 > 10",
  },
]
`;

exports[`Components GbInvestment Call 'addPromotionMessageTemplate' 2`] = `
Array [
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2,1 > 10",
  },
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
]
`;

exports[`Components GbInvestment Call 'addPromotionMessageTemplate' 3`] = `
Array [
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2,1 > 10",
  },
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
]
`;

exports[`Components GbInvestment Call 'removePromotionMessage' 1`] = `
Array [
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P1(124) P2(67) P3(19) P4(10)",
  },
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2,1 > 10",
  },
]
`;

exports[`Components GbInvestment Call 'removePromotionMessage' 2`] = `
Array [
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatory 9 < 4,3,2,1 > 10",
  },
]
`;

exports[`Components GbInvestment Change "lang" value 1`] = `
Array [
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 P1(124) P2(67) P3(19) P4(10)",
  },
  Object {
    "active": false,
    "message": "P1(124) P2(67) P3(19) P4(10) Observatoire 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 P4(10) P3(19) P2(67) P1(124)",
  },
  Object {
    "active": false,
    "message": "P4(10) P3(19) P2(67) P1(124) Observatoire 9 → 10",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 1 2 3 4",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 4 3 2 1",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 1,2,3,4",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 → 10 4,3,2,1",
  },
  Object {
    "active": false,
    "message": "Observatoire 9 < 4,3,2,1 > 10",
  },
]
`;

exports[`Components GbInvestment Change "result" value 1`] = `
Object {
  "promotion": Array [
    Object {
      "active": false,
      "message": "Observatory 9 → 10 P1(124) P2(67) P3(19) P4(10)",
    },
    Object {
      "active": false,
      "message": "P1(124) P2(67) P3(19) P4(10) Observatory 9 → 10",
    },
    Object {
      "active": false,
      "message": "Observatory 9 → 10 P4(10) P3(19) P2(67) P1(124)",
    },
    Object {
      "active": false,
      "message": "P4(10) P3(19) P2(67) P1(124) Observatory 9 → 10",
    },
    Object {
      "active": false,
      "message": "Observatory 9 → 10 1 2 3 4",
    },
    Object {
      "active": false,
      "message": "Observatory 9 → 10 4 3 2 1",
    },
    Object {
      "active": false,
      "message": "Observatory 9 → 10 1,2,3,4",
    },
    Object {
      "active": false,
      "message": "Observatory 9 → 10 4,3,2,1",
    },
    Object {
      "active": false,
      "message": "Observatory 9 → 10 4,3,2,1",
    },
    Object {
      "active": false,
      "message": "Observatory 9 → 10 4,3,2,1",
    },
    Object {
      "active": false,
      "message": "Observatory 9 < 4,3,2,1 > 10",
    },
  ],
  "result": Object {
    "cost": 650,
    "investment": Array [
      Object {
        "cumulativeInvestment": 526,
        "defaultParticipationIndex": -1,
        "expectedParticipation": 124,
        "isInvestorParticipation": false,
        "participation": 124,
        "preparation": 402,
        "reward": 65,
        "roi": 0,
        "snipe": Object {
          "fp": 325,
          "roi": -201,
        },
      },
      Object {
        "cumulativeInvestment": 593,
        "defaultParticipationIndex": -1,
        "expectedParticipation": 67,
        "isInvestorParticipation": false,
        "participation": 67,
        "preparation": 402,
        "reward": 35,
        "roi": 0,
        "snipe": Object {
          "fp": 325,
          "roi": -258,
        },
      },
      Object {
        "cumulativeInvestment": 631,
        "defaultParticipationIndex": -1,
        "expectedParticipation": 19,
        "isInvestorParticipation": false,
        "participation": 19,
        "preparation": 421,
        "reward": 10,
        "roi": 0,
        "snipe": Object {
          "fp": 325,
          "roi": -306,
        },
      },
      Object {
        "cumulativeInvestment": 641,
        "defaultParticipationIndex": -1,
        "expectedParticipation": 10,
        "isInvestorParticipation": false,
        "participation": 10,
        "preparation": 421,
        "reward": 5,
        "roi": 0,
        "snipe": Object {
          "fp": 325,
          "roi": -315,
        },
      },
      Object {
        "defaultParticipationIndex": -1,
        "expectedParticipation": 0,
        "isInvestorParticipation": false,
        "preparation": 430,
        "reward": 0,
        "roi": 0,
        "snipe": Object {
          "fp": 0,
          "roi": 0,
        },
      },
    ],
    "level": 10,
    "otherInvestment": Array [],
    "totalPreparations": 430,
  },
}
`;
